Fix encryption in IPAM ENI mode

aditighag opened the pull request:
The PR consists of two main fixes to make encryption functional on EKS - 

- Add routing rules for cilium_host ENI secondary IP address
- Attach `bpf_network` program to the correct interface for decryption to work

Following is out of scope of the PR, and will be addressed in follow-up PRs to simplify backporting :

- Refactor `loader.Reinitialize(...)`
- Review back-and-forth type conversions for `RoutingInfo`
- Writing datapath header defines (`IPV4_ENCRYPT_IFACE` and `ENCRYPT_IFACE`) 

**Release note**
```release-note
Make pod-to-pod encryption functional in the IPAM ENI mode.
```

aditighag committed:
ipam: Rename ENI Master field to PrimaryMAC

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
ipam/node: Introduce ENI info for cilium_host

This is a preparatory commit that introduces ENI related
information for cilium_host (aka router). Specifically,
the information includes primary ENI interface that the
secondary cilium_host IP is associated with, primary ENI MAC address
and VPC CIDRs.
Also, refactor the current IPAM internal IP allocating methods
so that the ENI information can be retrieved for cilium_host (aka router).

This information can be useful for setting up encryption
related state such as ip rules and interface to attach
BPF programs to.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
ipam: Add routes for cilium_host ENI address

Previously, we weren't adding policy based routing rules
for cilium_host secondary IP address. These rules ensure
that traffic from source node's cilium_host interface destined to the
remote node in the cluster VPC is routed through the primary
ENI interface for cilium_host. From there on, the traffic is routed to the
AWS infrastructure router, which in turn routes it to the destination node.

Fix involves setting up ip rules to the per ENI routing table.

Verified that the correct rule is added, and encrypted traffic is
reaching the destination node.

root@ip-192-168-38-245:/home/cilium# ip addr show cilium_host
5: cilium_host@cilium_net: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 9001 qdisc noqueue state UP group default qlen 1000
    link/ether da:61:19:32:0b:37 brd ff:ff:ff:ff:ff:ff
    inet 192.168.148.108/32 scope link cilium_host
       valid_lft forever preferred_lft forever
    inet6 fe80::d861:19ff:fe32:b37/64 scope link
       valid_lft forever preferred_lft forever
root@ip-192-168-38-245:/home/cilium# ip rule show | grep 192.168.148.108
111:	from 192.168.148.108 to 192.168.0.0/16 lookup 11

Tcpdump on the remote node shows encrypted packets getting received -

tcpdump -eni eth1 dst 192.168.142.218
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth1, link-type EN10MB (Ethernet), capture size 262144 bytes
17:32:52.216989 02:93:9c:00:a2:eb > 02:72:ef:86:3e:17, ethertype IPv4 (0x0800), length 130: 192.168.117.26 > 192.168.142.218: ESP(spi=0x00000003,seq=0x64), length 96
17:32:52.217694 02:93:9c:00:a2:eb > 02:72:ef:86:3e:17, ethertype IPv4 (0x0800), length 122: 192.168.117.26 > 192.168.142.218: ESP(spi=0x00000003,seq=0x65), length 88
17:32:52.217896 02:93:9c:00:a2:eb > 02:72:ef:86:3e:17, ethertype IPv4 (0x0800), length 206: 192.168.117.26 > 192.168.142.218: ESP(spi=0x00000003,seq=0x66), length 172

Signed-off-by: Aditi Ghag <aditi@cilium.io>

christarazi commented on daemon/cmd/ipam.go:
```diff
@@ -209,6 +209,12 @@ func (d *Daemon) allocateDatapathIPs(family datapath.NodeAddressingFamily) (rout
+			result.PrimaryMAC, result.InterfaceNumber, option.Config.EnableIPv4Masquerade)
+
+		node.SetRouterEniInfo(routingInfo)
+	}
```
Shouldn't this be gated on `option.Config.IPAM == ipamopt.IPAMENI`? 

I see that we are ignoring the error returned from `NewRoutingInfo` which is presumably being done because we don't want to error out if we aren't in ENI IPAM mode? I think it might be better to simply gate on the ENI IPAM mode and reinstate the error. This way we can alert if anything went wrong when setting up the router IP.

aditighag replied:
Good catch!

christarazi commented on pkg/node/address.go:
```diff
@@ -46,6 +47,7 @@ var (
 	ipv6NodePortAddrs map[string]net.IP // iface name => ip addr
 	ipv4AllocRange    *cidr.CIDR
 	ipv6AllocRange    *cidr.CIDR
+	routerEniInfo     *linuxrouting.RoutingInfo // Applicable in the ENI IPAM mode.
```
I wonder if we're better off simply adding the only the fields from the `RoutingInfo` struct that we need, rather than the whole struct. Mostly because it doesn't feel right for this package to be imported inside `pkg/node`.

For example, we can store:

- IPv4 Router CIDRs
- Interface number for ENI device
- MAC addr for ENI device

joestringer replied:
If the main concern is about package dependencies, then we could just define a routing info interface in this package and dependency-inject the routing info. That could be useful for unit-testing the new functionality (eg, make sure that the routing info is used in the right way in this package).

In practice, the above fields represent 3/5 fields in the `RoutingInfo` which is mostly pretty generic-looking, so I'm not sure how much finessing is required here; independently setting various fields which are all derived from the same source of truth doesn't seem like it will bring much benefit (rather it'll just be additional boilerplate).

christarazi replied:
Yeah, from a code readability point of view, it's not clear at all why the node package all of a sudden has a dependency on a thing that is only used in Azure or ENI mode. Consequently, having a variable name called `routerENIInfo`, which is the only reference to anything IPAM-specific in the node package, just seems like we're missing a piece of abstraction. Same issue with the heath endpoint in ENI mode with the daemon package as well.

This feels like something that should be addressed in a later refactor, but for now, I think going with an interface is a decent enough middle-ground.

christarazi replied:
Also the name `linuxrouting` seems to be a bad choice in hindsight. Yes it concerns itself with configuring Linux-specific routes and such, but it's only applicable for ENI and Azure. Naming is hard :/

aditighag replied:
>In practice, the above fields represent 3/5 fields in the RoutingInfo which is mostly pretty generic-looking, so I'm not sure how much finessing is required here

Correct. Also, even though I'm not familiar with AKS, it seems like the `gatewayIP` field is relevant, and could come into play in future.

re: package dependency

I 100% agree that the dependency seems out of place. Sure, interface based abstraction seems reasonable. Thanks!

christarazi commented on pkg/datapath/loader/base.go:
```diff
@@ -332,7 +352,7 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 	if option.Config.IPAM == ipamOption.IPAMENI {
 		var err error
-		if sysSettings, err = addENIRules(sysSettings); err != nil {
+		if sysSettings, err = addENIRules(sysSettings, nodeAddressing); err != nil {
```
Do we have to pass `nodeAddressing`? Can't we get away with just calling `node.GetInternalIPv4Router()` inside `addENIRules` which would eliminate the need to change the interface of both `Reinitialize()` and `addENIRules()`.

aditighag replied:
>Can't we get away with just calling node.GetInternalIPv4Router() inside

That's what I originally did, but decided to use the redirection via nodeAddressing in case we need to support IPv6 in this path (currently IPv6 doesn't seem to be supported with ENI IPAM mode though), and keep it consistent with the existing code.

Anyway I pushed a new commit couple of minutes back that gets rid off the additional argument.

christarazi reviewed (commented):
Nice! Core idea of the PR seems good to me. I have some questions on how we handle dependencies / importing packages.

christarazi commented on pkg/datapath/loader/base.go:
```diff
@@ -158,6 +160,24 @@ func addENIRules(sysSettings []sysctl.Setting) ([]sysctl.Setting, error) {
+	}
+	for _, cidr := range cidrs {
+		if err := route.ReplaceRule(route.Rule{
+			Priority: linux_defaults.RulePriorityEgressv2,
```
This will actually depend on whether `egress-multi-home-ip-rule-compat` is set or not. It defaults to false. True would mean that we want to maintain compatibility with the older ENI datapath (pre https://github.com/cilium/cilium/pull/14269)

joestringer replied:
Right, ideally these particular details (table /priority) should be handled in the linuxrouting package and this package should just ask for the table number (or just the routes? ... or just for the routes to be installed?).

Maybe this entire logic should just be pushed to a new function like `SetupRoutes()` exposed from `linuxrouting`? That way all of the logic around priority, table, RouterInfo introspection can be handled in just that one package rather than spreading it also to this package which must now take care of nuanced details of the route configuration.

christarazi replied:
That sounds like a reasonable idea to me. Much better than having seemingly random logic scattered around the codebase regarding rules and such.

aditighag replied:
>should be handled in the linuxrouting package 

Agreed. 

>This will actually depend on whether egress-multi-home-ip-rule-compat is set or not.

I don't think I follow this. AFAIU, this flag comes into play only during upgrade? But in this case, we were never installing the `cilium_host` rules until now?

joestringer replied:
The flag determines priority and table number for the rules, so this code should be consistent with which approach is used to determining these two values. The concern is if this isn't consistent, then the new rule which we install here may point to the wrong route table.

aditighag replied:
>The flag determines priority and table number for the rules, so this code should be consistent with which approach is used to determining these two values.

 Ah ok. Prior to v1.9 (or whichever release this flag was introduced in), the ip rules were installed in table Idx so we want to install the new rules to the same table Id.

joestringer replied:
FWIW the related PR is https://github.com/cilium/cilium/pull/14269 and it is still being backported. Both modes should be supported for now (until we deprecate & remove the flag).

joestringer commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -260,3 +260,24 @@ func retrieveIfIndexFromMAC(mac mac.MAC, mtu int) (index int, err error) {
+
+	links, err = netlink.LinkList()
+	if err != nil {
+		err = fmt.Errorf("unable to list interfaces: %s", err)
```
This needs to use `%w` to wrap the error properly.

joestringer commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -260,3 +260,24 @@ func retrieveIfIndexFromMAC(mac mac.MAC, mtu int) (index int, err error) {
+		}
+	}
+
+	err = fmt.Errorf("interface with MAC %s not found", mac)
```
ideally, errors should not include dynamic details like endpoint-specific state in them. I think we can move the `mac` piece into the `log.WithFields(...)` in the caller. This structured logging allows log handling software to better track & collapse log messages because if this error occurs for multiple MACs, then the error message will be consistent across all of them.

christarazi replied:
The nuance in this code though is that it'll be in the CNI code path, so the logs won't be in the agent, but rather in kubelet. Given that, I'd be inclined to staying with dynamic formatting. WDYT?

joestringer replied:
In this particular case, I think this is new functionality only called by the Cilium-agent code right?

I agree though if the error is going through cilium-cni out to kubelet then that's a reasonable argument to include dynamic details directly in the error messages.

christarazi replied:
Ha, you're totally right :sweat_smile:. I've had `pkg/datapath/linux/routing/routing.go` hardwired in my head to be called from CNI code

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -351,6 +351,15 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+		if info := node.GetRouterEniInfo(); info != nil {
+			iface, err := linuxrouting.RetrieveIfaceNameFromMAC(info.MasterIfMAC.String())
+			if err != nil {
+				log.WithError(err).Fatalf("failed to set encrypt interface in the ENI ipam mode")
```
This should use `Fatal()`, not `Fatalf()`. Also, I think we're supposed to capitalize the first word in log messages (Note, this is for log messages, not error messages; error messages should be all lower case).

joestringer replied:
(I also wondered about whether `Fatal` is appropriate here, but it looks like `Fatal` is already used at this point in the daemon so introducing one more usage in the case where we cannot fetch the MAC seems unlikely to incur additional friction for users)

joestringer reviewed (changes requested):
Just a few fairly minor organizational nits, core changes LGTM.

joestringer commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -255,6 +256,79 @@ func retrieveIfIndexFromMAC(mac mac.MAC, mtu int) (index int, err error) {
-func computeTableIDFromIfaceNumber(num int) int {
+// ComputeTableIDFromIfaceNumber returns a computed per-ENI route table ID for the given
+// ENI interface number.
+func ComputeTableIDFromIfaceNumber(num int) int {
```
Does this still need to be exported?

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -30,6 +31,7 @@ import (
 	"github.com/cilium/cilium/pkg/datapath/alignchecker"
 	"github.com/cilium/cilium/pkg/datapath/linux/linux_defaults"
 	"github.com/cilium/cilium/pkg/datapath/linux/route"
+	linuxrouting "github.com/cilium/cilium/pkg/datapath/linux/routing"
```
Looking at this again, now that we've shifted the actual route table specifics into the routing package I'm fine with this import. We clearly already directly import other packages under `datapath/linux` so if we want to improve the abstractions used in this area of the code more, then I think we should think about that outside of the context of this fix that we plan to backport.

joestringer reviewed (approved):
LGTM, thanks!

aditighag committed:
encryption/loader: Attach bpf_network program to correct interface for ENI mode

Problem: Auto detection of encryption.interface leads to bpf_network program
that does decryption to get attached to a wrong network interface.
As a result, packets were not decrypted on the destination remote node.

If users don't specify the cilium `encryption.interface`
configuration, the interface can be auto detected based on main routing
table. This breaks in ENI IPAM mode, where cilium_host is allocated an IP
address from the pool of secondary IP addresses of a primary ENI interface.
The IP allocation is influenced by the specified `first-interface-index` configuration.
Hence, retrieve the correct primary interface for cilium_host for bpf_network
program to be attached.

Fix is verified based on the xfrm stats -

Packet stats on source node :

src 192.168.117.26 dst 192.168.142.218
	proto esp spi 0x00000003(3) reqid 1(0x00000001) mode tunnel
	replay-window 0 seq 0x00000000 flag  (0x00000000)
	mark 0x3e00/0xff00 output-mark 0xe00
	aead rfc4106(gcm(aes)) 0xea5799388f441ba52b8c60c4f11c084bcdf5373e (160 bits) 128
	anti-replay context: seq 0x0, oseq 0x9, bitmap 0x00000000
	sel src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
	lifetime config:
	  limit: soft (INF)(bytes), hard (INF)(bytes)
	  limit: soft (INF)(packets), hard (INF)(packets)
	  expire add: soft 0(sec), hard 0(sec)
	  expire use: soft 0(sec), hard 0(sec)
	lifetime current:
	  540(bytes), 9(packets)
	  add 2021-02-05 18:12:08 use 2021-02-05 18:13:01
	stats:
	  replay-window 0 replay 0 failed 0

Packet stats on dest node :

ip -s xfrm s
src 192.168.142.218 dst 192.168.117.26
	proto esp spi 0x00000003(3) reqid 1(0x00000001) mode tunnel
	replay-window 0 seq 0x00000000 flag  (0x00000000)
	mark 0x3e00/0xff00 output-mark 0xe00
	aead rfc4106(gcm(aes)) 0xea5799388f441ba52b8c60c4f11c084bcdf5373e (160 bits) 128
	anti-replay context: seq 0x0, oseq 0x6, bitmap 0x00000000
	sel src 0.0.0.0/0 dst 0.0.0.0/0 uid 0
	lifetime config:
	  limit: soft (INF)(bytes), hard (INF)(bytes)
	  limit: soft (INF)(packets), hard (INF)(packets)
	  expire add: soft 0(sec), hard 0(sec)
	  expire use: soft 0(sec), hard 0(sec)
	lifetime current:
	  360(bytes), 6(packets)
	  add 2021-02-05 18:12:08 use 2021-02-05 23:19:07
	stats:
	  replay-window 0 replay 0 failed 0

Signed-off-by: Aditi Ghag <aditi@cilium.io>

christarazi commented on daemon/cmd/ipam.go:
```diff
@@ -209,6 +209,16 @@ func (d *Daemon) allocateDatapathIPs(family datapath.NodeAddressingFamily) (rout
+		routingInfo, err = linuxrouting.NewRoutingInfo(result.GatewayIP, result.CIDRs,
+			result.PrimaryMAC, result.InterfaceNumber, option.Config.EnableIPv4Masquerade)
+		if err != nil {
+			err = fmt.Errorf("failed to create router info %w", err)
```
`"failed to create router info: %w"` (colon)

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -195,6 +196,44 @@ func Delete(ip net.IP, compat bool) error {
 	return nil
 }
 
+// SetupRules installs routing rules based on the passed attributes. It accounts
+// for option.Config.EgressMultiHomeIPRuleCompat while configuring the rules.
```
Might make sense to clarify this to say routing rules for Azure / ENI modes

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -195,6 +196,44 @@ func Delete(ip net.IP, compat bool) error {
 
+// SetupRules installs routing rules based on the passed attributes. It accounts
+// for option.Config.EgressMultiHomeIPRuleCompat while configuring the rules.
+func SetupRules(from, to *net.IPNet, mac string, ifaceNum int) error {
```
Might also make sense to call this `SetupEgressRules`

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -195,6 +196,44 @@ func Delete(ip net.IP, compat bool) error {
+		prio = linux_defaults.RulePriorityEgressv2
+		tableId = computeTableIDFromIfaceNumber(ifaceNum)
+	}
+	return route.ReplaceRule(route.Rule{
+		Priority: prio,
+		From:     from,
+		To:       to,
+		Table:    tableId,
+	})
+}
```
You could now use this function to replace where `Configure()` creates egress rules (just a few lines above)

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -195,6 +196,44 @@ func Delete(ip net.IP, compat bool) error {
+		tableId int
+	)
+
+	if option.Config.EgressMultiHomeIPRuleCompat {
```
If we converge the egress rule creation with `Configure()`, then we can leverage the fact that `Configure()` already takes in a `compat` bool which represents this option, so we can get rid of this line and replace it with:

```
if compat {
```

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -255,6 +294,41 @@ func retrieveIfIndexFromMAC(mac mac.MAC, mtu int) (index int, err error) {
+func retrieveIfaceIdxFromMAC(mac string) (int, error) {
+	iface, err := retrieveIfaceFromMAC(mac)
+	if err != nil {
+		err = fmt.Errorf("failed to get iface index with MAC %w", err)
```
`"failed to get iface index with MAC %w"` ->

`"failed to get iface index by MAC: %w"`

christarazi commented on pkg/datapath/loader/base.go:
```diff
@@ -158,6 +160,19 @@ func addENIRules(sysSettings []sysctl.Setting) ([]sysctl.Setting, error) {
+	info := node.GetRouterInfo()
+	cidrs := info.GetIPv4CIDRs()
+	routerIP := net.IPNet{
+		IP:   nodeAddressing.IPv4().Router(),
```
Were you able to call `node.GetInternalIP()` to get rid of the additional param to this function?

christarazi commented on pkg/datapath/loader/base.go:
```diff
@@ -331,8 +346,18 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+			mac := info.GetMac()
+			iface, err := linuxrouting.RetrieveIfaceNameFromMAC(mac.String())
+			if err != nil {
+				log.WithError(err).WithField("mac", mac).Fatal("Failed to set encrypt interface in the ENI ipam mode")
```
`logfields.MAC`

christarazi reviewed (commented):
LGTM, thanks! Mostly minor nits to followup on; not urgent
